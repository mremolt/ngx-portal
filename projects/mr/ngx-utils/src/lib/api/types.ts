import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

export interface Constructor<T> {
  new (...args: any[]): T;
}

export interface ApiTokenSelector {
  select(state: any): string;
}

export interface ApiTokenSelectorConstructor {
  new (): ApiTokenSelector;
}

export interface ApiStartAction extends Action {
  readonly payload?: any;
}

export interface ApiSuccessAction<T> extends Action {
  readonly payload: T;
}

export interface ApiErrorAction extends Action {
  readonly payload: HttpErrorResponse;
}

export type ApiActions<T> = Readonly<{
  start: Constructor<ApiStartAction>;
  success: Constructor<ApiSuccessAction<T>>;
  error: Constructor<ApiErrorAction>;
  complete?: Constructor<Action>;
}>;

export type ApiRequestConfig<T> = Readonly<{
  actions: ApiActions<T>;
  request: HttpRequest<T>;
}>;
