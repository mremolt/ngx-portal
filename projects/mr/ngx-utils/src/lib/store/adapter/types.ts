import { DeepReadonly } from 'utility-types';

export type NormalizedEntities<Entity extends object> = DeepReadonly<{
  collection: { [key: string]: Entity };
  [key: string]: { [key: string]: object };
}>;

export type CollectionState<Entity extends object> = DeepReadonly<{
  entities: NormalizedEntities<Entity>;
  result: string[];
  selected: string | null;
  loading: boolean;
  loaded: boolean;
  dirty: boolean;
  updatedAt: Date | null;
}>;
