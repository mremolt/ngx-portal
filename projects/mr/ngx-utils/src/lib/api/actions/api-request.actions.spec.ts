import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

import { ApiErrorAction, ApiStartAction, ApiSuccessAction } from '../types';
import {
  ApiDeleteAction,
  ApiError,
  ApiGetAction,
  ApiPostAction,
  ApiPutAction,
  ApiRequest,
} from './api-request.actions';

export class FetchUsersStart implements ApiStartAction {
  public readonly type = '[Users] Fetch Start';
  constructor(public payload: any) {}
}

export class FetchUsersSuccess<T> implements ApiSuccessAction<T> {
  public readonly type = '[Users] Fetch Success';
  constructor(public payload: T) {}
}

export class FetchUsersError implements ApiErrorAction {
  public readonly type = '[Users] Fetch Error';
  constructor(public payload: HttpErrorResponse) {}
}

export class FetchUsersComplete implements Action {
  public readonly type = '[Users] Fetch Complete';
}

describe('ApiRequest Actions', () => {
  describe('ApiRequest', () => {
    it('builds the correct action', () => {
      const action = new ApiRequest({
        request: new HttpRequest('GET', 'users'),
        actions: {
          start: FetchUsersStart,
          success: FetchUsersSuccess,
          error: FetchUsersError,
          complete: FetchUsersComplete,
        },
      });
      expect(action).toBeTruthy();
      expect(action).toMatchSnapshot();
    });
  });

  describe('ApiError', () => {
    it('builds the correct action', () => {
      const action = new ApiError(new HttpErrorResponse({}));
      expect(action).toBeTruthy();
      expect(action).toMatchSnapshot();
    });
  });

  describe('ApiGetAction', () => {
    it('builds the correct action', () => {
      const action = new ApiGetAction('users', {
        start: FetchUsersStart,
        success: FetchUsersSuccess,
        error: FetchUsersError,
        complete: FetchUsersComplete,
      });
      expect(action).toBeTruthy();
      expect(action).toMatchSnapshot();
    });
  });

  describe('ApiPostAction', () => {
    it('builds the correct action', () => {
      const action = new ApiPostAction(
        'users',
        { id: 42 },
        {
          start: FetchUsersStart,
          success: FetchUsersSuccess,
          error: FetchUsersError,
          complete: FetchUsersComplete,
        }
      );
      expect(action).toBeTruthy();
      expect(action).toMatchSnapshot();
    });
  });

  describe('ApiPutAction', () => {
    it('builds the correct action', () => {
      const action = new ApiPutAction(
        'users',
        { id: 42 },
        {
          start: FetchUsersStart,
          success: FetchUsersSuccess,
          error: FetchUsersError,
          complete: FetchUsersComplete,
        }
      );
      expect(action).toBeTruthy();
      expect(action).toMatchSnapshot();
    });
  });

  describe('ApiDeleteAction', () => {
    it('builds the correct action', () => {
      const action = new ApiDeleteAction('users', {
        start: FetchUsersStart,
        success: FetchUsersSuccess,
        error: FetchUsersError,
        complete: FetchUsersComplete,
      });
      expect(action).toBeTruthy();
      expect(action).toMatchSnapshot();
    });
  });
});
