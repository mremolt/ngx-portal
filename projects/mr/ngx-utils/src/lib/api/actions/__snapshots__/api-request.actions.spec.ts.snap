// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiRequest Actions ApiDeleteAction builds the correct action 1`] = `
ApiDeleteAction {
  "payload": Object {
    "actions": Object {
      "complete": [Function],
      "error": [Function],
      "start": [Function],
      "success": [Function],
    },
    "request": HttpRequest {
      "body": null,
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "method": "DELETE",
      "params": HttpParams {
        "cloneFrom": null,
        "encoder": HttpUrlEncodingCodec {},
        "map": null,
        "updates": null,
      },
      "reportProgress": false,
      "responseType": "json",
      "url": "users",
      "urlWithParams": "users",
      "withCredentials": false,
    },
  },
  "type": "[API] Request",
}
`;

exports[`ApiRequest Actions ApiError builds the correct action 1`] = `
ApiError {
  "payload": HttpErrorResponse {
    "error": null,
    "headers": HttpHeaders {
      "headers": Map {},
      "lazyUpdate": null,
      "normalizedNames": Map {},
    },
    "message": "Http failure response for (unknown url): undefined undefined",
    "name": "HttpErrorResponse",
    "ok": false,
    "status": 0,
    "statusText": "Unknown Error",
    "url": null,
  },
  "type": "[API] Error",
}
`;

exports[`ApiRequest Actions ApiGetAction builds the correct action 1`] = `
ApiGetAction {
  "payload": Object {
    "actions": Object {
      "complete": [Function],
      "error": [Function],
      "start": [Function],
      "success": [Function],
    },
    "request": HttpRequest {
      "body": null,
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "method": "GET",
      "params": HttpParams {
        "cloneFrom": null,
        "encoder": HttpUrlEncodingCodec {},
        "map": null,
        "updates": null,
      },
      "reportProgress": false,
      "responseType": "json",
      "url": "users",
      "urlWithParams": "users",
      "withCredentials": false,
    },
  },
  "type": "[API] Request",
}
`;

exports[`ApiRequest Actions ApiPostAction builds the correct action 1`] = `
ApiPostAction {
  "payload": Object {
    "actions": Object {
      "complete": [Function],
      "error": [Function],
      "start": [Function],
      "success": [Function],
    },
    "request": HttpRequest {
      "body": Object {
        "id": 42,
      },
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "method": "POST",
      "params": HttpParams {
        "cloneFrom": null,
        "encoder": HttpUrlEncodingCodec {},
        "map": null,
        "updates": null,
      },
      "reportProgress": false,
      "responseType": "json",
      "url": "users",
      "urlWithParams": "users",
      "withCredentials": false,
    },
  },
  "type": "[API] Request",
}
`;

exports[`ApiRequest Actions ApiPutAction builds the correct action 1`] = `
ApiPutAction {
  "payload": Object {
    "actions": Object {
      "complete": [Function],
      "error": [Function],
      "start": [Function],
      "success": [Function],
    },
    "request": HttpRequest {
      "body": Object {
        "id": 42,
      },
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "method": "PUT",
      "params": HttpParams {
        "cloneFrom": null,
        "encoder": HttpUrlEncodingCodec {},
        "map": null,
        "updates": null,
      },
      "reportProgress": false,
      "responseType": "json",
      "url": "users",
      "urlWithParams": "users",
      "withCredentials": false,
    },
  },
  "type": "[API] Request",
}
`;

exports[`ApiRequest Actions ApiRequest builds the correct action 1`] = `
ApiRequest {
  "payload": Object {
    "actions": Object {
      "complete": [Function],
      "error": [Function],
      "start": [Function],
      "success": [Function],
    },
    "request": HttpRequest {
      "body": null,
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "method": "GET",
      "params": HttpParams {
        "cloneFrom": null,
        "encoder": HttpUrlEncodingCodec {},
        "map": null,
        "updates": null,
      },
      "reportProgress": false,
      "responseType": "json",
      "url": "users",
      "urlWithParams": "users",
      "withCredentials": false,
    },
  },
  "type": "[API] Request",
}
`;
